### DISCUSSION TITLE
How can I compare two multipage TIFFs for differences?

### DISCUSSION THREAD

**Discussion by chrispy-snps** [ImageMagick/ImageMagick#3279] [Help] [answered]
Hello everyone! I'd like to visually compare two PDF files for differences. I can use Ghostscript to create two multi-page TIFF files from PDF files:

```
gs -dNOPAUSE  -dBATCH -sDEVICE=tiff24nc -sOutputFile='pdf1.tiff' -sCompression=lzw -dUseCropBox -r72 pdf1.pdf
gs -dNOPAUSE  -dBATCH -sDEVICE=tiff24nc -sOutputFile='pdf2.tiff' -sCompression=lzw -dUseCropBox -r72 pdf2.pdf
```

Now I need some way to compare these two multipage TIFFs and highlight differences to the user.

Although there are some scattered posts about multipage TIFF comparison with ImageMagick ([1](https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=7815)) ([2](https://stackoverflow.com/questions/52380568/difference-of-multipage-tiff)), I'm unable to get the posted commands to work for me (and indeed, other posters reply that it doesn't work for them either.)

So instead of listing all the commands that don't work, I'd like to start fresh and ask what _is_ the proper command to compare two multipage TIFFs and somehow highlight the differences for the user?

I'm using Ubuntu 18.04 LTS, whose latest ImageMagick version is 6.9.7-4. If there are improvements in this area, I can certainly acquire/compile the latest version.

Thanks in advance!

 - Chris

**Comment by snibgo** [comment]
Don't bother yet about the multi-page aspect. First, address this:

> ... what is the proper command to compare two multipage TIFFs and somehow highlight the differences for the user?

There are many possibilities, depending on what outputs you want.

I work chiefly with photos, and I want to know what pixels have changed, and by how much.
```
magick in1.tiff[0] in2.tiff[0] -compose Difference -composite -colorspace Gray -auto-level out.tiff
```
Pixel with the greatest differences will be white, and pixels with no differences will be black. "-colorspace Gray" and "-auto-level" are both optional. 

But your need may be different. You might care only about what pixels are different, irrespective of how different they are. And you might want to see a faint image of one of the inputs, perhaps in grayscale, with different pixels show as bright red. Or whatever.

**Comment by fmw42** [comment]
What is the content of your pages? Are they images or are they text?

**Reply by chrispy-snps** [reply]
@fmw42 - good question! These are technical documentation manuals, consisting of mostly text, some tables/lists, figures (vector drawings or screen captures). There are no photographs. For example,

![pdf1](https://user-images.githubusercontent.com/50950969/108573597-0dd2e600-72e3-11eb-9640-df605c31fdf1.png)

The PDF publishing software we use (Syncrosoft's PDF Chemistry) is configured by CSS. It has a default CSS style, which we customize with our own CSS settings. As new releases of their software update their CSS settings, we sometimes need to update our CSS to compensate. The differences are typically minor changes in spacing/positioning, font size, and other such minor CSS-oriented differences.

My goal is to set up a regression test that compares ~100 PDF files to check for differences between the current known-good publishing engine and the new publishing engine.

@snibgo - this is part of what I'm hoping this group can help me with! The differences are typically in content layout, where the content has either shifted, or a piece of content has appeared/disappeared.

The primary need is to identify changed versus unchanged pages by scripting. It doesn't need to be pretty to look at - I'll end up just opening the actual PDFs anyway.

If your operation yields solid black pages for identical pages, then I should be able to apply some kind of trim/crop operation to differentiate between pages with and without changes. I'll play with the command you suggested and see what I can come up with! 

If you'd like to experiment too, here are two 2-page TIFF files with a minor layout difference:

[two_tiffs.zip](https://github.com/ImageMagick/ImageMagick/files/6013694/two_tiffs.zip)

**Reply by fmw42** [reply]
Were these scanned or are they actual vector text? If text, then you can extract from the PDF and compare the words. If scanned, then you likely need OCR to convert to vector text and then compare.  ImageMagick is a raster processor. Any rotation, scaling, etc will cause differences that may not relate to the text information differences.

**Reply by chrispy-snps** [reply]
@fmw42 - it's actual text. However, I'm looking for visual differences - things moving by a few pixels, a font accidentally changing from nonitalic to italic, things like that - rather than text differences. (Sometimes a text label might appear/disappear depending on what CSS property changed, but that will also result in a visual difference.)

Basically I'm publishing two CSS-formatted PDF files in two different versions of a publishing program, and I need to make sure the PDFs are visually identical between them.

**Reply by fmw42** [reply]
ImageMagick has a compare tool that can be used in two different ways. 1) if the images are aligned and the same size, it can show where there are differences (like subtracting the two) or 2) if one is smaller than the other, it will move the one over the other to to find the best match location and then give you the differences.  However, it does not work well if there are rotations or scale differences. See 

https://imagemagick.org/Usage/compare/
https://imagemagick.org/script/compare.php

For multipage tiffs, you will need to write a script loop over each pair of pages to do the comparison.  The compare tool only works on one pair of pages at a time.

**Comment by snibgo** [comment, marked as answer]
I suggest:
```
magick ^
  pdf1.tiff null: pdf2.tiff ^
  -background None ^
  -compose Difference -layers Composite ^
  +write x.tiff ^
  -format "IsDifferent=%%[fx:maxima==0?0:1]\n" ^
  info:
```
This makes x.tiff with two images. Each images shows differences. In addition, there is text output with one line per output image:
```
IsDifferent=1
IsDifferent=1
```
This shows "1" when the output has any differences (ie any non-black pixels), or "0" when there are no differences (ie all pixels are black).

**Reply by chrispy-snps** [reply]
@snibgo - with your considerable help, I was able to create the following PDF comparison utility:

https://github.com/chrispy-snps/compare-pdf-images

I have wanted to write this for years. Thank you for giving me the *magick* to make it work!

**Reply by snibgo** [reply]
Good stuff. Thanks for sharing the script.
